matrix/matrix.py

```diff
@@ -0,0 +1,198 @@
+import operator
+from itertools import repeat
+
+
+class Matrix:
+    """An immutable object representing a matrix data-type. This
+    implementation only supports even-side matrices (n * n).
+
+    Attributes:
+        _matrix (tuple): contains the data about the matrix. Each line of
+                         the matrix is represented by a tuple.
+    """
+    __slots__ = '_matrix'
+
+    def __init__(self, *lines):
+        if not self._is_valid_data(*lines):
+            raise ValueError('Data is not valid')
+
+        super(Matrix, self).__setattr__('_matrix', tuple(lines))
+
+    def __setattr__(self, key, value):
+        """Defines that Matrix attributes are immutable."""
+        raise AttributeError(f'{self.__class__} instance cannot be modified')
+
+    @staticmethod
```

Comment:
Usage of `staticmethod` is highly rare and is usually not recommended
If your IDE yells about the unused `self`, you can either ignore it or define this function outside of class's scope